optimum/torch_export_patches/_treenode_registry.py

```diff
@@ -0,0 +1,287 @@
+
+    # DynamicCache serialization is different in transformers and does not
+    # play way with torch.export.export.
+    # see test test_export_dynamic_cache_cat with NOBYPASS=1
+    # :: NOBYBASS=1 python _unittests/ut_torch_export_patches/test_dynamic_class.py -k e_c
+    # This is caused by this line:
+    # torch.fx._pytree.register_pytree_flatten_spec(
+    #           DynamicCache, _flatten_dynamic_cache_for_fx)
+    # so we remove it anyway
+    # BaseModelOutput serialization is incomplete.
+    # It does not include dynamic shapes mapping.
+    for cls, version in wrong.items():
+        if (
+            cls in torch.utils._pytree.SUPPORTED_NODES
+            and cls not in PATCH_OF_PATCHES
+            # and pv.Version(torch.__version__) < pv.Version("2.7")
+            and (version is None or pv.Version(transformers.__version__) >= pv.Version(version))
+        ):
+            assert cls in registration_functions, (
+                f"{cls} has no registration functions mapped to it, "
+                f"available options are {list(registration_functions)}"
+            )
+            if verbose:
+                print(f"[_fix_registration] {cls.__name__} is unregistered and registered first")
+            unregister_class_serialization(cls, verbose=verbose)
+            registration_functions[cls](verbose=verbose)  # type: ignore[arg-type, call-arg]
+            if verbose:
+                print(f"[_fix_registration] {cls.__name__} done.")
+            # To avoid doing it multiple times.
+            PATCH_OF_PATCHES.add(cls)
+
+    # classes with no registration at all.
+    done = {}
+    for k, v in registration_functions.items():
+        done[k] = v(verbose=verbose)  # type: ignore[arg-type, call-arg]
+    return done
+
+
+def serialization_functions(
+    patch_transformers: bool = False, patch_diffusers: bool = False, verbose: int = 0
+) -> dict[type, Callable[[int], bool]]:
+    """Returns the list of serialization functions."""
+    supported_classes: set[type] = set()
+    classes: dict[type, Callable[[int], bool]] = {}
+    all_functions: dict[type, str | None] = {}
+
+    if patch_transformers:
+        from optimum.torch_export_patches._treenode import (
+            SUPPORTED_DATACLASSES,
+            flatten_dynamic_cache,
+            flatten_encoder_decoder_cache,
+            flatten_mamba_cache,
+            flatten_sliding_window_cache,
+            flatten_static_cache,
+            flatten_with_keys_dynamic_cache,
+            flatten_with_keys_encoder_decoder_cache,
+            flatten_with_keys_mamba_cache,
+            flatten_with_keys_sliding_window_cache,
+            flatten_with_keys_static_cache,
+            unflatten_dynamic_cache,
+            unflatten_encoder_decoder_cache,
+            unflatten_mamba_cache,
+            unflatten_sliding_window_cache,
+            unflatten_static_cache,
+        )
+
+        from .serialization.transformers_impl import (
+            __dict__ as dtr,
```

Comment:
@xadupre I forgot to ask what is this

Reply:
dtr contains the list of functions in one submodule, then the function looks into all of the them to see if it is a patch or not.

Reply:
Is it possible to register explicitly?

Reply:
We can. The patches need to be removed after the export if a user wants to use it to compute discrepancies. So I chose to restore everything back to its original state including serialization functions.